core/src/main/scala/cats/syntax/either.scala

```diff
@@ -1,13 +1,163 @@
 package cats
 package syntax
 
-import cats.data.Xor
+import cats.data.{EitherT, Ior, Validated, ValidatedNel, Xor}
+import scala.reflect.ClassTag
+import scala.util.{Failure, Success, Try}
 
 trait EitherSyntax {
   implicit def catsSyntaxEither[A, B](eab: Either[A, B]): EitherOps[A, B] = new EitherOps(eab)
+
+  implicit def catsSyntaxEitherObject(either: Either.type ): EitherObjectOps = new EitherObjectOps(either)
 }
 
 final class EitherOps[A, B](val eab: Either[A, B]) extends AnyVal {
+  def foreach(f: B => Unit): Unit = eab.fold(_ => (), f)
+
+  def getOrElse[BB >: B](default: => BB): BB = eab.fold(_ => default, identity)
+
+  def orElse[C, BB >: B](fallback: => Either[C, BB]): Either[C, BB]= eab match {
+    case Left(_)      => fallback
+    case r @ Right(_) => r.asInstanceOf[Either[C, BB]]
+  }
+
+  def recover[BB >: B](pf: PartialFunction[A, BB]): Either[A, BB] = eab match {
+    case Left(a) if pf.isDefinedAt(a) => Right(pf(a))
+    case _                            => eab
+  }
+
+  def recoverWith[AA >: A, BB >: B](pf: PartialFunction[A, Either[AA, BB]]): Either[AA, BB] = eab match {
+    case Left(a) if pf.isDefinedAt(a) => pf(a)
+    case _                            => eab
+  }
+
+  def valueOr[BB >: B](f: A => BB): BB = eab.fold(f, identity)
+
+  def forall(f: B => Boolean): Boolean = eab.fold(_ => true, f)
+
+  def exists(f: B => Boolean): Boolean = eab.fold(_ => false, f)
+
+  def ensure[AA >: A](onFailure: => AA)(f: B => Boolean): Either[AA, B] =
+    eab.fold(_ => eab, b => if (f(b)) eab else Left(onFailure))
+
+  def toIor: A Ior B = eab.fold(Ior.left, Ior.right)
+
+  def toOption: Option[B] = eab.fold(_ => None, Some(_))
+
+  def toList: List[B] = eab.fold(_ => Nil, _ :: Nil)
+
+  def toTry(implicit ev: A <:< Throwable): Try[B] = eab.fold(a => Failure(ev(a)), Success(_))
+
+  def toValidated: Validated[A, B] = eab.fold(Validated.Invalid.apply, Validated.Valid.apply)
+
+  /** Returns a [[ValidatedNel]] representation of this disjunction with the `Left` value
+   * as a single element on the `Invalid` side of the [[NonEmptyList]]. */
+  def toValidatedNel[AA >: A]: ValidatedNel[AA, B] = eab.fold(Validated.invalidNel, Validated.valid)
+
+  def withValidated[AA, BB](f: Validated[A, B] => Validated[AA, BB]): Either[AA, BB] =
+    f(toValidated).toEither
+
+  def to[F[_], BB >: B](implicit F: Alternative[F]): F[BB] =
+    eab.fold(_ => F.empty, F.pure)
+
+  def bimap[C, D](fa: A => C, fb: B => D): Either[C, D] = eab match {
+    case Left(a)  => Left(fa(a))
+    case Right(b) => Right(fb(b))
+  }
+
+  def map[C](f: B => C): Either[A, C] = eab match {
+    case l @ Left(_) => l.asInstanceOf[Either[A, C]]
```

Comment:
Can we make some casting methods so we can more easily check safety? Something like:

```scala
def castLeft[A, B, C](left: Left[A, B]): Left[A, C] = ...
```
Similar with right?